### PR TITLE
drawpile: add build options and split into multiple top-level attrs

### PR DESCRIPTION
**Pull request by fgaz**
###### Motivation for this change

Right now the `drawpile` derivation builds a full package, but someone may want only the server or client. This pr makes it possible and adds some drawpile variants to `all-packages.nix`.

Is the way I did it ok? Am I unnecessarily polluting the global namespace?

Depends on #57568.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/graphics/drawpile/default.nix [hunk 1]
```diff
@@ -25,24 +34,47 @@ stdenv.mkDerivation rec {
...
+    optionals enableKisTablet  kisDeps                           ;
+  commonDeps = [
     cmake
-    qtbase qtsvg qtmultimedia qttools
     karchive
-    # optional deps:
-    #   server-specific:
-    libsodium # ext-auth support
+  ];
+  clientDeps = [
```
Comment (veprbl) [inline pkgs/applications/graphics/drawpile/default.nix]: You unconditionally pass `clientDeps`, `serverDeps`, `kisDeps`, `serverOptionalDeps` and `clientOptionalDeps` as attributes to derivation which effectively makes all of the possible dependencies required to be built before building this derivation.

File: pkgs/applications/graphics/drawpile/default.nix [hunk 2]
```diff
@@ -25,24 +34,47 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [
     extra-cmake-modules
   ];
-  buildInputs = [
-    # common deps:
+  buildInputs =
+    commonDeps ++
+    optionals buildClient     (clientDeps ++ clientOptionalDeps) ++
```
Comment (veprbl) [inline pkgs/applications/graphics/drawpile/default.nix]: What is the purpose of splitting `clientOptionalDeps` if they are not optional.

File: pkgs/applications/graphics/drawpile/default.nix [hunk 3]
```diff
@@ -25,24 +34,47 @@ stdenv.mkDerivation rec {
...
+      ${optionalString (!buildClient    )  "-DCLIENT=off"   } \
+      ${optionalString (!buildServer    )  "-DSERVER=off"   } \
+      ${optionalString (!buildServerGui )  "-DSERVERGUI=off"} \
+      ${optionalString ( buildExtraTools)  "-DTOOLS=on"     } \
+      ${optionalString ( enableKisTablet)  "-DKIS_TABLET=on"} \
+      -DCMAKE_INSTALL_PREFIX=$out
+  '';
 
-  meta = with stdenv.lib; {
+  meta = {
```
Comment (veprbl) [inline pkgs/applications/graphics/drawpile/default.nix]: `meta.platforms` is missing

**Comment by veprbl** [comment]
@fgaz What problem are you solving with this? Adding extra variants means burden in maintaining more packages.

File: pkgs/applications/graphics/drawpile/default.nix (hunk 1 above)
Comment (fgaz) [reply pkgs/applications/graphics/drawpile/default.nix]: Oops, yeah this should be outside of that set, thanks.

**Comment by fgaz** [comment]
@veprbl Thanks for the review!

> What problem are you solving with this

I want to separate client and server, so people can install for example just the server part without bringing in GUI/client deps, but yeah, I got carried away and added `-headless` and `-kis`.

What about:
* keeping all the options as arguments
* adding two extra args to enable/disable the optional deps
  * or even an arg for each optional dep (`enableGifSupport` etc)
* just adding a `-server-headless` variant so hydra can build it

File: pkgs/applications/graphics/drawpile/default.nix (hunk 2 above)
Comment (fgaz) [reply pkgs/applications/graphics/drawpile/default.nix]: Mostly just laziness :-P

Please see https://github.com/NixOS/nixpkgs/pull/57573#issuecomment-475859339

**Comment by veprbl** [comment]
> adding two extra args to enable/disable the optional deps

I'm not convinced there is such a big value in having knobs to fine tune all these optional dependencies. I think we usually enable all optional stuff unconditionally for maximal convenience. If someone doesn't want to build with `giflib` they can always use `drawpile.override { giflib = null; }`. You can mention what dependencies are optional in a comment, this should be more than enough for an advanced user.

> just adding a -server-headless variant so hydra can build it

This sounds good. If you expect that most people would rather use either a full gui client and server or a headless server install then it greatly reduces number of options that we need to provide. We can keep `-kis` variant if you think you need it.

**Comment by fgaz** [comment]
Waiting for https://github.com/drawpile/Drawpile/issues/744 for `kis`. I'll just modify the default without adding any `-kis` variant if they say to enable it.

**Comment by veprbl** [comment]
It seems like they say that they don't need it. Also you probably don't need to package any special setups unless you use them yourself.

**Comment by fgaz** [comment]
All done. Ping @veprbl

File: pkgs/applications/graphics/drawpile/default.nix [hunk 4]
```diff
@@ -1,48 +1,89 @@
...
-    libvpx # WebM video export
+    extra-cmake-modules
   ];
-  configurePhase = "cmake -DCMAKE_INSTALL_PREFIX=$out .";
+  buildInputs =
+    commonDeps ++
+    optionals buildClient      clientDeps ++
+    optionals buildServer      serverDeps ++
+    optionals enableKisTablet  kisDeps    ;
+  configurePhase = ''
```
Comment (veprbl) [inline pkgs/applications/graphics/drawpile/default.nix]: Please don't override `configurePhase`, use `cmakeFlags` instead.

**Comment by veprbl** [comment]
@fgaz Thanks!

File: pkgs/applications/graphics/drawpile/default.nix (hunk 4 above)
Comment (fgaz) [reply pkgs/applications/graphics/drawpile/default.nix]: Thanks for the fixup!

**Pull request merged**